[Network] #59 - Onboarding API 연결

joho2022 opened the pull request:
## 🌴 작업한 브랜치
- #59


## ✅ 작업한 내용
<!-- 작업한 내용 적어주세요! -->
- [x] 서비스 이용약관 정보 조회 API
- [x] 닉네임 검증 API
- [x] 닉네임 등록 API
- [x] 선호장르 목록 조회 API
- [x] 선호장르 검색 API
- [x] 선호장르 등록 API

<!--
```
넣고싶은 코드가 있다면 적어주세요
```
-->


## ❗️PR Point
<!-- 부족했던 점 혹은 개선하고 싶은 방향이 있다면 얘기해주세요 -->

<!--
```
넣고싶은 코드가 있다면 적어주세요
```
-->


## 📸 스크린샷
<!-- 스크린 샷을 첨부해주세요 -->
|기능|스크린샷|
|:--:|:--:|
|약관조회|<img src = "https://github.com/user-attachments/assets/5d74020e-0b2e-4d35-9d60-9af5b56e90c2" width ="450">|
|닉네임등록|<img src = "https://github.com/user-attachments/assets/7989b1fb-ea0f-4a17-9d19-41a44f495c35" width ="450">|
|선호장르등록|<img src = "https://github.com/user-attachments/assets/d3f3b186-e0db-411c-95ef-c63ddd4f6109" width ="450">|


## 📟 관련 이슈
- Resolved: #59

joho2022 committed:
[Feat] #59 - 온보딩 약관 동의 API 연결

joho2022 committed:
[Feat] #59 - 닉네임 검증, 등록 구현 및 API 연결

joho2022 committed:
[Feat] #59 - 선호 장르 조회, 검색 등록 구현 및 API 연결

joho2022 committed:
[Chore] #59 - TODO 주석 삭제

Johyerin commented on Napzak-iOS/Napzak-iOS/Presentation/Onboarding/View/GenreSelectionView.swift:
```diff
@@ -143,8 +151,11 @@ extension GenreSelectionView {
             Button {
-                // TODO: 유저 장르 등록
-                authRouter.push(next: .completed)
+                Task {
+                    if await viewModel.registerSelectedGenres() {
+                        authRouter.push(next: .completed)
+                    }
+                }
```
P1

서버 연동 이전엔 확인을 못했는데 스크롤 가장 아래로 내렸을 때 버튼과 장르 목록 뷰가 겹치는 현상이 있어서요! 패딩값 적용해주셔야 할 것 같아유~~

![image](https://github.com/user-attachments/assets/a156e29c-a688-4209-95f3-bacddc9356aa)

Johyerin commented on Napzak-iOS/Napzak-iOS/Presentation/Onboarding/ViewModel/GenreSelectionViewModel.swift:
```diff
@@ -6,51 +6,94 @@
     init() {
-        loadDefaultGenres()
+        $searchText
+            .debounce(for: 0.5, scheduler: RunLoop.main)
+            .removeDuplicates()
+            .sink { [weak self] newSearchText in
+                Task {
+                    await self?.searchGenres()
+                }
```
옹 디바운스 👍🏻

OneTen19 replied:
요게 그 저번에 말했던, 계속 API 요청 보내는 게 아니라 몇 초 간격을 두고 보내는 그건가!?

Johyerin reviewed (approved):
LGTM 깔끔하네요!! 연결된 서버 기능 잘 동작하는 것 확인했습니다~ 고생핑

OneTen19 commented on Napzak-iOS/Napzak-iOS/Global/Components/UsernameInputField.swift:
```diff
@@ -60,58 +69,67 @@ struct UsernameInputField: View {
-        let validPattern = "^(?=.*[가-힣]|[A-Za-z0-9])[가-힣A-Za-z0-9]{2,}$"
-        let consonantVowelPattern = "^[ㄱ-ㅎㅏ-ㅣ]+$"
+    private func containsIncompleteHangul(_ text: String) -> Bool {
+        let chosung = "ㄱㄲㄴㄷㄸㄹㅁㅂㅃㅅㅆㅇㅈㅉㅊㅋㅌㅍㅎ"
+        let jungsung = "ㅏㅐㅑㅒㅓㅔㅕㅖㅗㅘㅙㅚㅛㅜㅝㅞㅟㅠㅡㅢㅣ"
```
초성 정성맨..

OneTen19 commented on Napzak-iOS/Napzak-iOS/Network/Base/ErrorResponseDTO.swift:
```diff
@@ -0,0 +1,13 @@
+
+import Foundation
+
+struct ErrorResponseDTO: Decodable {
+  let status: Int
+  let message: String
+}
```
근데 이거도  BaseResponseDTO에 
```
let status
let message
let data?
```
이렇게 있던데 data는 nullable이니까 typealias였나 그거 활용하면 안되는건가여??

OneTen19 commented on Napzak-iOS/Napzak-iOS/Presentation/Onboarding/View/GenreSelectionView.swift:
```diff
@@ -53,6 +53,14 @@ struct GenreSelectionView: View {
                 isSearchFocused = false
             }
         }
+        .task {
+            await viewModel.fetchAllGenres()
+        }
```
궁금한게 이런 방식으로 task를 사용하면 어느 시점에 task 내부 코드들이 실행되는건가요?
onTapGesture나 onChange같은 이벤트를 관측하는 부분 내에 들어있는 것도 아닌데 fetchAllGenres가 언제 실행되는건지...?
뷰가 보여질때 실행되는건가...?

OneTen19 commented on Napzak-iOS/Napzak-iOS/Presentation/Onboarding/ViewModel/GenreSelectionViewModel.swift:
```diff
@@ -6,51 +6,94 @@
     
     init() {
-        loadDefaultGenres()
+        $searchText
+            .debounce(for: 0.5, scheduler: RunLoop.main)
+            .removeDuplicates()
+            .sink { [weak self] newSearchText in
+                Task {
+                    await self?.searchGenres()
+                }
+            }
+            .store(in: &cancellables)
```
바인딩도 아닌데 searchText에 왜 $가 있지..했는데
맨 밑에 store 있는 거 보니 혹시 요게 컴바인 문법인가요!?

OneTen19 reviewed (approved):
수고하셨습니다!!

lalaurrel reviewed (approved):
수고하셨습니다

joho2022 committed:
[Chore] #59 - 스크롤 뷰 하단 여백 추가

joho2022 committed:
[Merge] #59 - Conflicts 해결

# Conflicts:
#	Napzak-iOS/Napzak-iOS.xcodeproj/project.pbxproj
#	Napzak-iOS/Napzak-iOS/Presentation/ProfileEdit/ProfileEditView.swift
